### ISSUE TITLE
Weird portafolio weights and final portfolio value

### ISSUE THREAD

**Issue by Human-HLW** [CaioSBC/RLPortfolio#7]
Hi!  

I found I can't train the model very well so I selected only one asset and observed abnormal portafolio weights and Final portfolio value. Did I miss something?

1# epoch
![Image](https://github.com/user-attachments/assets/e0e96ac3-f1d2-4481-9568-806b197c8885)
or
![Image](https://github.com/user-attachments/assets/d810937d-0eb3-4f1b-850e-412863f9f1b7)
(They are in the opposite direction)
or
![Image](https://github.com/user-attachments/assets/5344d2f0-7472-4c17-87a8-2767d3f9ffbd)

5# epoch
![Image](https://github.com/user-attachments/assets/c67dc1a2-acba-4ee6-ab4c-ef9ae2ecca3f)

10# epoch
![Image](https://github.com/user-attachments/assets/9fbbe511-134f-4043-937d-f74920a072f5)

15# epoch
![Image](https://github.com/user-attachments/assets/90abe2ef-5c55-41ec-b34c-89a81d2f76da)

20# epoch
![Image](https://github.com/user-attachments/assets/b32c6337-6cc3-4521-b28f-03d44f22c237)

The Final portfolio value is decreasing
![Image](https://github.com/user-attachments/assets/99eaaa5f-7a8a-469d-a72e-113c9a2d98e3)

[train_data.csv](https://github.com/user-attachments/files/19214636/train_data.csv)

[test_data.csv](https://github.com/user-attachments/files/19214638/test_data.csv)

import pandas as pd
from rlportfolio.environment import PortfolioOptimizationEnv

def convert_date_format(date_str):
    if '/' in date_str:
        return date_str.replace('/', '-')
    else:
        return date_str

df_train = pd.read_csv("train_data.csv")
df_train = df_train.dropna()
df_train['date'] = df_train['date'].apply(convert_date_format)
df_train['date'] = pd.to_datetime(df_train['date'])
print("date", df_train['date'])
print("df_train =", df_train)

environment = PortfolioOptimizationEnv(
        df_train, # data to be used
        100000    # initial value of the portfolio
    )

from rlportfolio.algorithm import PolicyGradient

algorithm = PolicyGradient(environment)

algorithm.train(20)

**Comment by CaioSBC** [comment]
Hello!

I downloaded your data and made a few experiments. The issue is related to hyperparameters and the lack of normalization.

### About normalization

There are two ways of normalize data in this library. The first one is to utilize the `PortfolioOptimizationEnv` (check [this link)](Península Way) and several articles in the research field use the state normalization method. However, I have noticed that it is usually more consistent to just normalize the data before using it like this:

```python
from rlportfolio.data import GroupByScaler

# normalize data
scaler = GroupByScaler(by="tic", columns=["high", "low", "close"])
df_train = scaler.fit_transform(df_train)
```

As you can see, the library has a "scikit-learn-like" scaler which normalizes the price columns of the input data dividing it by its max value.

### About the hyperparameters

When I instantiated the algorithm, I only changed one hyperparameter:

```python
algorithm = PolicyGradient(environment, sample_bias=0.001, use_tensorboard=True)
```

The `sample_bias` is an important hyperparameter. It is basically the probability of a geometric distribution which samples the batch of data to perform the optimization. Currently, the default value is 1 (I will probably change that), so that only the most recent batch of data is optimized. The bigger the `sample_bias`, the more the algorithm will focus on more recent data points. Since your training data has 1900+ data points, I used 0.001.

I also set `use_tensorboard` to True in order to be able to visualize the training process.

### Training

I trained the algorithm for 40000 steps using the following code:

```python
algorithm.train(40000, logging_period=1000)
```

I set `logging_period` to 1000 so that every 1000 training steps, the performance of the agent is evaluated using a full training episode. The following results were achieved:

![Image](https://github.com/user-attachments/assets/2b92a692-5ea3-4f48-a42a-6825dbe9f6fa)

![Image](https://github.com/user-attachments/assets/aeb50882-86a4-4eee-b2e5-2658e9507cd8)

![Image](https://github.com/user-attachments/assets/56953469-68c1-4a25-aaf8-595fb1f38421)

![Image](https://github.com/user-attachments/assets/7407f91e-39b4-46d6-a613-272ad930624e)

As you can see in figure 2, the agent effective learns a policy (he knows when to sell all the stocks and when to buy them) and it leads him to achieve profit (1200% cumulative return as seen in figure 1). Note that it doesn't mean that the agent will perform well in the test data: this algorithm suffers with overfitting (this is something I am investigating), but you can, eventually, generate a good policy.

Additionally, figure 3 shows the FAPV of the portfolio growing during the training. The FAPV is a metric called "final accumulative portfolio value" and is basically the last value of the portfolio divided by its initial value. FInally, figure 4 demonstrates the loss function value is increasing, proving that the optimization is being done correctly.

Let me know if you need any further help.
